src/openms/include/OpenMS/ANALYSIS/ID/AhoCorasickDA.h

```diff
@@ -0,0 +1,437 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Chris Bielow$
+// $Authors: Patricia Scheil$
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/CHEMISTRY/AASequence.h>
+#include <OpenMS/CONCEPT/Types.h>
+#include <vector>
+#include <unordered_map>
+#include <stdint.h>
+
+
+namespace OpenMS
+{
+
+  /**
+   * @brief Extended Aho-Corasick algorithm capable of matching ambiguous amino acids in the pattern (i.e. proteins).
+   * The trie of the Aho Corasick automaton is based on a compact double-array (DA) :
+   *
+   * Susumu  Yata  Masaki  Oono  Toru  Sumitomo  Kazuhiro  Morita,  Masao  Fuketa  and  Jun’ichiAoe.
+   * A  compact  static  double-array  keeping  character  codes.
+   * Information Processing and Managemnet : 237-247 (2007)
+   */
+  class OPENMS_DLLAPI AhoCorasickDA
+  {
+
+  public:
+
+    /// Constructor support DNA, RNA and Protein sequences (accept lower and capital letter except ambiguous aa (B/J/X/Z))
+    explicit AhoCorasickDA(const std::vector<String>& seq) : sequences_(seq)
+    {
+      construct_();
+    }
+
+
+    /// Destructor
+    ~AhoCorasickDA() = default;
+
+
+    /// Set new protein sequence. All previous data is forgotten.
+    void setProtein(const char* prot, const uint8_t amb_max = 3);
+
+
+    /**
+     * @brief search for and numerate all occurrences of the peptides in the protein
+     * @param[out] pos_in_protein Position (zero-based) in the protein
+     * @param[out] peptide_index index of sequence in peptide database
+     * @return False if end of protein is reached. True if a hit is found.
+     */
+    bool findNext(Size& pos_in_protein, Size& peptide_index);
+
+
+    // for debugging
+    bool retrievalCDA();
+
+    // for debugging
+    void printDA(bool arrays);
+
+    UInt32 stat();
+
+
+  private:
+
+    /// Node informations needed for the construction of the BC array. Contains informations about used nodes, used base values and empty elements.
+    struct BuildInformation_
+    {
+      explicit BuildInformation_(const UInt32 index) : next(index+1), used_flag(0), prev(index-1),  base_flag(0), check(0){};
+      UInt32 next        :31 ;    ///< next empty node in DA
+      UInt32 used_flag   :1  ;    ///< node is used
+      UInt32 prev        :31 ;    ///< previous empty node in the DA
+      UInt32 base_flag   :1  ;    ///< this value is used as base
+      UInt32 check           ;    ///< parent node TODO not necessary
+    };
+
+    /// The properties of a node in the base-check-array (BC array)
+    struct BCNode_
+    {
+      UInt32 base       :22 ;    ///< base value of the node
+      UInt32 lcheck     :8  ;    ///< arc label from parent node. Only code of the aa is stored
```

Comment:
it **might** be faster if you switch the order of `base` and `lcheck` because that should(!) save some bitshifting when accessing `lcheck` .. but it might not help....
don't redo the benchmarks!